### ISSUE TITLE
1520.5.0 stable channel update crashes systemd-resolved when using PRIVATE_IPV4 as DNS server

### ISSUE THREAD

**Issue by spunon** [coreos/bugs#2190]
# Issue Report #

## Bug ##
In the newest stable update that just came out, we're seeing systemd-resolved fail after the node reboots, and on subsequent reboots
```
$ app-9 user # cat /etc/resolv.conf
cat: /etc/resolv.conf: No such file or directory
$ app-9 user # systemctl status systemd-resolved
● systemd-resolved.service - Network Name Resolution
   Loaded: loaded (/usr/lib/systemd/system/systemd-resolved.service; enabled; vendor preset: disabled)
   Active: failed (Result: exit-code) since Thu 2017-10-12 07:50:25 UTC; 12s ago
     Docs: man:systemd-resolved.service(8)
           https://www.freedesktop.org/wiki/Software/systemd/resolved
           https://www.freedesktop.org/wiki/Software/systemd/writing-network-configuration-managers
           https://www.freedesktop.org/wiki/Software/systemd/writing-resolver-clients
  Process: 3199 ExecStart=/usr/lib/systemd/systemd-resolved (code=exited, status=226/NAMESPACE)
 Main PID: 3199 (code=exited, status=226/NAMESPACE)
      CPU: 51ms

Oct 12 07:50:25 app-9 systemd[1]: systemd-resolved.service: Unit entered failed state.
Oct 12 07:50:25 app-9 systemd[1]: systemd-resolved.service: Failed with result 'exit-code'.
Oct 12 07:50:25 app-9 systemd[1]: systemd-resolved.service: Service has no hold-off time, scheduling restart.
Oct 12 07:50:25 app-9 systemd[1]: Stopped Network Name Resolution.
Oct 12 07:50:25 app-9 systemd[1]: systemd-resolved.service: Start request repeated too quickly.
Oct 12 07:50:25 app-9 systemd[1]: Failed to start Network Name Resolution.
Oct 12 07:50:25 app-9 systemd[1]: systemd-resolved.service: Unit entered failed state.
Oct 12 07:50:25 app-9 systemd[1]: systemd-resolved.service: Failed with result 'exit-code'.

$ app-9 user # systemctl cat systemd-resolved
# /usr/lib/systemd/system/systemd-resolved.service
#  This file is part of systemd.
#
#  systemd is free software; you can redistribute it and/or modify it
#  under the terms of the GNU Lesser General Public License as published by
#  the Free Software Foundation; either version 2.1 of the License, or
#  (at your option) any later version.

[Unit]
Description=Network Name Resolution
Documentation=man:systemd-resolved.service(8)
Documentation=https://www.freedesktop.org/wiki/Software/systemd/resolved
Documentation=https://www.freedesktop.org/wiki/Software/systemd/writing-network-configuration-managers
Documentation=https://www.freedesktop.org/wiki/Software/systemd/writing-resolver-clients
After=systemd-networkd.service network.target
Before=network-online.target nss-lookup.target
Wants=nss-lookup.target

# On kdbus systems we pull in the busname explicitly, because it
# carries policy that allows the daemon to acquire its name.
Wants=org.freedesktop.resolve1.busname
After=org.freedesktop.resolve1.busname

[Service]
Type=notify
Restart=always
RestartSec=0
ExecStart=/usr/lib/systemd/systemd-resolved
WatchdogSec=3min
CapabilityBoundingSet=CAP_SETUID CAP_SETGID CAP_SETPCAP CAP_CHOWN CAP_DAC_OVERRIDE CAP_FOWNER CAP_NET_RAW CAP_NET_BIND_SERVICE
PrivateTmp=yes
PrivateDevices=yes
ProtectSystem=strict
ProtectHome=yes
ProtectControlGroups=yes
ProtectKernelTunables=yes
ProtectKernelModules=yes
MemoryDenyWriteExecute=yes
RestrictRealtime=yes
RestrictAddressFamilies=AF_UNIX AF_NETLINK AF_INET AF_INET6
SystemCallFilter=~@clock @cpu-emulation @debug @keyring @module @mount @obsolete @raw-io @reboot @swap
SystemCallArchitectures=native
ReadWritePaths=/run/systemd

[Install]
WantedBy=multi-user.target
Alias=dbus-org.freedesktop.resolve1.service
```
### Container Linux Version ###

```
$ cat /etc/os-release
NAME="Container Linux by CoreOS"
ID=coreos
VERSION=1520.5.0
VERSION_ID=1520.5.0
BUILD_ID=2017-10-10-2231
PRETTY_NAME="Container Linux by CoreOS 1520.5.0 (Ladybug)"
ANSI_COLOR="38;5;75"
HOME_URL="https://coreos.com/"
BUG_REPORT_URL="https://issues.coreos.com"
COREOS_BOARD="amd64-usr"
...
BUG_REPORT_URL="https://issues.coreos.com"
```

### Environment ###

AWS, we've seen this exact behavior on several instance types

### Reproduction Steps ###

  1. Install the stable version before (1465.8.0) from ami, with standard configuration for etcd2 and fleet, with an automatic update/reboot strategy
  2. Write a conf file in /etc/systemd/resolved.conf.d/
  3. Restart sytemd-resolved
  2. Update to 1520.5.0

**Comment by euank** [comment]
Can you clarify what configuration file you're writing in `/etc/systemd/resolved.conf.d/` in step 2 of your reproduction, @spunon?

The full logs of resolved (`journalctl -u systemd-resolved --no-pager`) could also be helpful.

**Comment by spunon** [comment]
00-consul-dns.conf : 
```
[Resolve]
DNS=${COREOS_PRIVATE_IPV4}
```

Journal logs from systemd-resolved:
```
Oct 12 17:47:33 app-9 systemd[1]: Starting Network Name Resolution...
Oct 12 17:47:33 app-9 systemd-resolved[784]: Positive Trust Anchors:
Oct 12 17:47:33 app-9 systemd-resolved[784]: . IN DS 19036 8 2 49aac11d7b6f6446702e54a1607371607a1a41855200fd2ce1cdde32f24e8fb5
Oct 12 17:47:33 app-9 systemd-resolved[784]: . IN DS 20326 8 2 e06d44b80b8f1d39a95c0b0d7c65d08458e880409bbc683457104237c7f8ec8d
Oct 12 17:47:33 app-9 systemd-resolved[784]: Negative trust anchors: 10.in-addr.arpa 16.172.in-addr.arpa 17.172.in-addr.arpa 18.172.in-addr.arpa 19.172.in-addr.arpa 20.172.in-addr.arpa 21.172.in-addr.arpa 22.172.in-addr.arpa 23.172.in-add
Oct 12 17:47:33 app-9 systemd-resolved[784]: Using system hostname 'app-9'.
Oct 12 17:47:33 app-9 systemd[1]: Started Network Name Resolution.
Oct 12 17:48:09 app-9 systemd[1]: Stopping Network Name Resolution...
Oct 12 17:48:09 app-9 systemd[1]: Stopped Network Name Resolution.
Oct 12 17:48:09 app-9 systemd[1]: Starting Network Name Resolution...
Oct 12 17:48:09 app-9 systemd[1]: systemd-resolved.service: Main process exited, code=exited, status=226/NAMESPACE
Oct 12 17:48:09 app-9 systemd[1]: Failed to start Network Name Resolution.
Oct 12 17:48:09 app-9 systemd[1]: systemd-resolved.service: Unit entered failed state.
Oct 12 17:48:09 app-9 systemd[1]: systemd-resolved.service: Failed with result 'exit-code'.
Oct 12 17:48:09 app-9 systemd[1]: systemd-resolved.service: Service has no hold-off time, scheduling restart.
Oct 12 17:48:09 app-9 systemd[1]: Stopped Network Name Resolution.
```

**Comment by spunon** [comment]
Updated the title to be more specific of my setup.

**Comment by bgilbert** [comment]
I don't have a consul setup at hand, but I couldn't immediately reproduce this by setting the DNS server using a similar drop-in.  `status=226/NAMESPACE` suggests that the additional security options added to `systemd-resolved.service` since CL 1465 may be causing the problem.  Here's the service file diff:

```diff
--- resolved-old	2017-10-12 14:43:51.736745913 -0700
+++ resolved-new	2017-10-12 14:44:06.660778843 -0700
@@ -9,10 +9,12 @@
 [Unit]
 Description=Network Name Resolution
 Documentation=man:systemd-resolved.service(8)
-Documentation=http://www.freedesktop.org/wiki/Software/systemd/resolved
-Documentation=http://www.freedesktop.org/wiki/Software/systemd/writing-network-configuration-managers
-Documentation=http://www.freedesktop.org/wiki/Software/systemd/writing-resolver-clients
+Documentation=https://www.freedesktop.org/wiki/Software/systemd/resolved
+Documentation=https://www.freedesktop.org/wiki/Software/systemd/writing-network-configuration-managers
+Documentation=https://www.freedesktop.org/wiki/Software/systemd/writing-resolver-clients
 After=systemd-networkd.service network.target
+Before=network-online.target nss-lookup.target
+Wants=nss-lookup.target
 
 # On kdbus systems we pull in the busname explicitly, because it
 # carries policy that allows the daemon to acquire its name.
@@ -28,14 +30,17 @@
 CapabilityBoundingSet=CAP_SETUID CAP_SETGID CAP_SETPCAP CAP_CHOWN CAP_DAC_OVERRIDE CAP_FOWNER CAP_NET_RAW CAP_NET_BIND_SERVICE
 PrivateTmp=yes
 PrivateDevices=yes
-ProtectSystem=full
+ProtectSystem=strict
 ProtectHome=yes
 ProtectControlGroups=yes
 ProtectKernelTunables=yes
+ProtectKernelModules=yes
 MemoryDenyWriteExecute=yes
 RestrictRealtime=yes
 RestrictAddressFamilies=AF_UNIX AF_NETLINK AF_INET AF_INET6
-SystemCallFilter=~@clock @cpu-emulation @debug @keyring @module @mount @obsolete @raw-io
+SystemCallFilter=~@clock @cpu-emulation @debug @keyring @module @mount @obsolete @raw-io @reboot @swap
+SystemCallArchitectures=native
+ReadWritePaths=/run/systemd
 
 [Install]
 WantedBy=multi-user.target
```

You could try using a drop-in to disable some of those additional options, or reverting back to the previous service file entirely, and see if that fixes your issue.  If you can provide detailed instructions to reproduce the issue, that would also be very helpful.

**Comment by ajeddeloh** [comment]
It's almost certainly the `ProtectSystem=strict` change that's causing this, although I haven't been able to reproduce this locally. Try disabling that option with a drop in.

I'm not sure why that would cause this problem, but it looks like that `apply_mount_namespace` is failing here: https://github.com/coreos/systemd/blob/v234-coreos/src/core/execute.c#L2626 since that's the only place that will return `EXIT_NAMESPACE`. I'm looking into why that might fail.

**Comment by spunon** [comment]
Also if it helps, here's our cloud-config:
```
#cloud-config

hostname: app-9
manage_etc_hosts: "localhost"
coreos:
  update:
    reboot-strategy: "etcd-lock"
  fleet:
    public-ip: $private_ipv4
    etcd_servers: http://$private_ipv4:2379
    metadata: "name=app-9,role=app,consul-agent=true"
  etcd2:
    proxy: on
    listen-client-urls: "http://0.0.0.0:2379"
    initial-cluster: infra0=http://10.0.7.8:2380,infra1=http://10.0.7.80:2380,infra2=http://10.0.7.180:2380
  units:
    - name: etcd2.service
      command: start
    - name: fleet.service
      command: start
```

**Comment by ajeddeloh** [comment]
I still haven't been able to reproduce it. https://github.com/coreos/bugs/issues/2193 looks related and adding this dropin fixed it for them:
```
[Service]
ProtectSystem=full
```

They also mentioned only some nodes failed; are you seeing all nodes fail or only some of them?

**Comment by ajeddeloh** [comment]
@spunon This actually looks like a kernel bug (see the upstream systemd issue if you want more details). Are you running nfs? It looks like it might be an nfs bug.

**Comment by bgilbert** [comment]
We're reverting to `ProtectSystem=full` pending a complete fix in upstream systemd. This should avoid the problem starting in alpha 1590.1.0, beta 1576.3.0, and stable 1520.9.0, which are due shortly.

Thanks for reporting.  The long-term fix for this issue will be tracked in #2193.
